std_execution.bs

```diff
@@ -669,6 +669,58 @@ Sender consumers could be made pipeable, but we have chosen to not do so.
+
+## Many senders can be trivially made awaitable ## {#design-senders-are-awaitable}
+
+If you choose to implement your sender-based algorithms as coroutines, you'll run into the issue of how to retrieve results from a passed-in sender. This is not a problem. If the coroutine type opts in to sender support -- trivial with the `execution::with_awaitable_senders` utility -- then a large class of senders are transparently awaitable from within the coroutine.
+
+For example, consider the following trivial implementation of the sender-based `retry` algorithm:
+
+<pre highlight="c++">
+template &lt;class S>
+  requires <i>single-typed-sender</i>&lt;S&> // See <a href="#spec-execution.coro_utils.as_awaitable">[execution.coro_utils.as_awaitable]</a>
+task&lt;<i>single-sender-value-type</i>&lt;S>> retry(S& s) {
+  for (;;) {
+    try {
+      co_return co_await s;
+    } catch(...) {
+    }
+  }
+}
+</pre>
+
+Only *some* senders can be made awaitable directly because of the fact that callbacks are more expressive than coroutines. An awaitable expression has a single type: the result value of the async operation. In contrast, a callback can accept multiple arguments as the result of an operation. What's more, the callback can have overloaded function call signatures that take different sets of arguments. There is no way to automatically map such senders into awaitables. The `with_awaitable_senders` utility recognizes as awaitables those senders that send a single value of a single type. To await another kind of sender, a user would have to first map its value channel into a single value of a single type -- say, with the `into_variant` sender algorithm -- before `co_await`-ing that sender.
+
+## Cancellation of a sender can unwind a stack of coroutines ## {#design-native-coro-unwind}
+
+When looking at the sender-based `retry` algorithm in the previous section, we can see that the value and error cases are correctly handled. But what about cancellation? What happens to a coroutine that is suspended awaiting a sender that completes by calling `execution::set_done`?
+
+When your task type's promise inherits from `with_awaitable_senders`, what happens is this: the coroutine behaves as if an *uncatchable exception* had been thrown from the `co_await` expression. (It is not really an exception, but it's helpful to think of it that way.) Provided that the promise types of the calling coroutines also inherit from `with_awaitable_senders`, or more generally implement a member function called `unhandled_done`, the exception unwinds the chain of coroutines as if an exception were thrown except that it bypasses `catch(...)` clauses.
+
+In order to "catch" this uncatchable done exception, one of the calling coroutines in the stack would have to await a sender that maps the done channel into either a value or an error. That is achievable with either the `execution::let_done` or `execution::upon_done` sender adaptors.
```

Comment:
I'm not sure that we should be recommending `let_done` here.
It's use within a coroutine is complicated as having the `let_done()` return a `just(std::nullopt_t)` will result in the overall sender having multiple result-types and then needing to be piped through `into_variant`.

It would be nicer if we could describe an `done_as_optional` algorithm and show an example of its use...
```c++
std::optional<T> result = co_await done_as_optional(some_operation(args...));
if (!result) {
  // completed with set_done()
} else {
  // completed with set_value()
  auto& value = *result;
}
```

Whereas doing the same thing with `let_done()` is more cumbersome:
```c++
std::optional<T> result = co_await transform(
  let_done(some_operation(args...), []() noexcept { return just(std::nullopt); }),
  [](auto&& value) -> std::optional<T> { return static_cast<decltype(value)>(value); });
```

Maybe also throw in an example of using `done_as_error<E>()`?
```c++
try {
  T result = co_await done_as_error<operation_cancelled>(some_operation(args...));
} catch (operation_cancelled) {
  ...
}
```
Or alternatively show that you can just let this exception propagate up through the call-stack of coroutines...